include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -332,39 +427,134 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
+      // extract relevant information from distributed triangulation
+      auto const construction_data =
+        TriangulationDescription::Utilities::create_description_from_triangulation(tria_copy,
+                                                                                   mpi_comm);
 
-      // create empty (fully distributed) triangulation
-      auto new_tria = std::make_shared<parallel::fullydistributed::Triangulation<dim, spacedim>>(
-        fine_triangulation->get_communicator());
+      // create fully distributed triangulation
+      auto level_tria =
+        std::make_shared<parallel::fullydistributed::Triangulation<dim, spacedim>>(mpi_comm);
 
       for(auto const i : fine_triangulation->get_manifold_ids())
         if(i != numbers::flat_manifold_id)
-          new_tria->set_manifold(i, fine_triangulation->get_manifold(i));
+          level_tria->set_manifold(i, fine_triangulation->get_manifold(i));
+
+      level_tria->create_triangulation(construction_data);
+
+      coarse_grid_triangulations[level - 1] = level_tria;
+
+      if(level > 1)
+        tria_copy.coarsen_global();
+    }
+
+    // TODO: The following code is a brute-force attempt to create a new
+    // partitioning of the mesh to be fed to a
+    // parallel::fullydistributed::Triangulation with fewer MPI processes. The
+    // main idea is to replicate the complete triangulation obtained in the
+    // previous loop on specific MPI processes and partition it from there
+    // again.
+    //
+    // Step 1: Since the full triangulation with refinements can in general be
+    // many GB of size, we only want to fit a single copy on each node. We
+    // identify the nodes by letting MPI count the number of MPI processes in
+    // a shared memory region.
+    unsigned int const n_processes_per_node = [&]() {
+      int rank;
+      MPI_Comm_rank(mpi_comm, &rank);
+
+      MPI_Comm comm_shared;
+      MPI_Comm_split_type(mpi_comm, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL, &comm_shared);
+
+      int size_shared;
+      MPI_Comm_size(comm_shared, &size_shared);
+      MPI_Comm_free(&comm_shared);
+
+      AssertThrow(size_shared == Utilities::MPI::max(size_shared, mpi_comm),
+                  ExcMessage("The mesh coarsening with shared memory only works if all nodes "
+                             "are populated with the same number of MPI ranks!"));
+      return size_shared;
+    }();
+
+
+    // Step 2: Create coarse grid on the first MPI process of each node
+    Triangulation<dim, spacedim> serial_tria;
+    replicate_distributed_triangulation_on_node(tria_copy,
+                                                mpi_comm,
+                                                n_processes_per_node,
+                                                serial_tria);
+
+    // Step 3: Continue as above but with the serial triangulation that gets
+    // distributed
+    unsigned int n_partitions = Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD);
```

Comment:
why `MPI_COMM_WORLD` (which we generally do not want to use in such files in ExaDG as far as I know)?

Reply:
Silly mistake. Thanks.